### ISSUE TITLE
Generic questions about WHDL Autoloader and overriding of parameters

### ISSUE THREAD

**Issue by Voljega** [midwan/amiberry#491]
Hello everyone,

Recalbox is now using Amiberry WHDL Autoloader.

While the compatibility is really good and much better than the previous methods, I have a few questions for which I'm not really sure of the response.

- What parameters exactly does the WHDL Autoloader overrides in amiberry during launch ? For a logical point of view and from what I understand most of the parameters should be set by the Autoloader : inputs, everything related to systems / platform used, graphic configuration ? 
I ask that because I have the feeling not everything is working perfectly in Recalbox current implementation : for instance CD32 whdl games do not use a CD32 joystick but a regular one, which seems odd.
Also graphical resolution doesn't seem to be ever changed by the AutoLoader, so some games are missing part of the image

- If the loader overrides parameters, is the GUI always correctly updated ? It might explain some discrepencies between uae for the launch of the game / override by Autoloader / actual behaviour of the game

TLDR : what part of the original uae parameters are overriden by the WHDL Autoloader once it boots and ignored ?

**Comment by Voljega** [comment]
hmmm anyone ? @HoraceAndTheSpider I assume you should know ? ;)

**Comment by HoraceAndTheSpider** [comment]
The autoloader only uses existing settings functions, so there certainly should not be disparity with what is shown in the GUI.

So initially it determines ecs/Aga/cd32 and uses the QuickStart functions to do this. 

This would override any uaeconfig default parameters.

To set the other prefs which are in the Xml, It basically manually parses commands as they would be specified in the .uae file .... unless there is a uae file in which case that is used instead.

You’ll need to check the wiki here to see what hardware setting get set by the xml, but it’s not many - chipset changes etc for compatibility and memory are the big ones, plus of course mounting the HDs.

Lastly (I think) it sets anything required which is host specific (I.e hostprefs) by the same method.

If something like cd32 controls are missing... do you still have Cd32 in the file name? Perhaps on recallbox the name parsing is not performing correctly (I.e it’s an OS difference)

**Comment by Voljega** [comment]
So I guess the issue might be that we generate an uae and not launch the lha of the game directly or something ?

Because CD32 is totally in the lha name (standard whdl lha are used)

Couldn't find in the wiki a page listing precisely what parameters the loader uses/sets, do you have a link ?

I also couldn't find the hostprefs file in our installation, shoukd it be manually created and where ?

Lastly if I read your comment right, graphical resolution is not set by the whdl autoloader ?

**Comment by Voljega** [comment]
@HoraceAndTheSpider So I was able to better investigate.

First lha games are launched with the command line `/usr/bin/amiberry -config=/tmp/amiga/conf/uaeconfig.uae`

Amiberry's WHDL AutoBooter is clearly used but something is amiss
I tested two games, for each of them you can find a link to a zip containing :
- captures of the GUI
- conf directort of the game
- retroarchcustom.cfg mentioned in the conf (not sure if it is really used )

**Cybernoid_v1.3_1088 :**
conf for the game in the whdl xml used in our install is :
```

**Comment by HoraceAndTheSpider** [comment]
Config= is not enough to trigger the autoloader, but possibly this is being ignored anyway if you are pointing at the default Config, which will never working together with the autoloader 

You need to use autoload= and point to the lha

**Comment by Voljega** [comment]
@HoraceAndTheSpider 

OK I replaced launching line by using autowhdload : 
`/usr/bin/amiberry', '-autowhdload=/recalbox/share/roms/amiga1200/[ShootEmUp]/Cybernoid_v1.3_1088.lha`

Staying on the same two examples, it's better as specific parameters of the whdl db seems to be taken into account, but there are still a lot of issues

First whdl db conf specifies SCREEN_HEIGHT of 200 for both games, which translate to 640*200 in GUI and it doesn't seem to be the right screen size for both game ? Especially Cybernoid misses a lot of the picture that way, I included screen caps in 200 and in more to show the difference.

Controls is also better, joystick type are setted correctly, but PORT0 device always stays as mouse for both game.
Custom controls are again totally and utterly false... see hotkey set as none and absolutely no button beeing set.

And I have two joysticks connected, configured and on before launching the game ....

Caps for Cybernoid : https://imgur.com/a/e10YnEB
Caps for Chaos Engine : https://imgur.com/a/SMPAyYP

**Comment by midwan** [comment]
I believe the `hostprefs.conf` can have global settings such as preferred controllers, that will be used with the WHDLoad booter. Have you checked there?

Also, please use `-autoload=` instead of `-autowhdload=`. They both do the same thing, but the second one is there for compatibility reasons only and might be removed at some point in the future. ;-)

**Comment by Voljega** [comment]
Thanks Dimitris for the response !

I tried looking for the hostprefs.conf file but actually couldn'tfind it anywhere on Recalbox.

And thecontrollers GUI issue also appears with ADF where this file hostprefs.conf isn't takzn into account anyway I guess ?

Coincidentally, the same issues with controllers GUI part also appears on Batocera distribution, but I didn't really investigate code there, suffice to say that vothe implementation are different anyway...

**Comment by midwan** [comment]
@Voljega 
It should be under the `whdboot` directory, did you check there?

Take a look at the relevant wiki article also: https://github.com/midwan/amiberry/wiki/WHDLoad-Auto-booting

**Comment by Voljega** [comment]
So it seems this file was indeed missing, creating it with values of the wiki 
```
CONTROLLER_1=joy1
CONTROLLER_2=joy2
;CONTROLLER_3=joy0
;CONTROLLER_4=joy3
CONTROLLER_MOUSE_1=mouse
CONTROLLER_MOUSE_2=joy0
;RETROARCH_QUIT=TRUE
;RETROARCH_MENU=TRUE
;RETROARCH_RESET=TRUE
;KEY_FOR_QUIT=F12
;KEY_FOR_MENU=F11
;DEADZONE=33
;STEREO_SPLIT=1
;SOUND_ON=FALSE
;ASPECT_RATIO_FIX=TRUE
;FRAMESKIP=TRUE
```
partially solves one of the issue : "Controls is also better, joystick type are setted correctly, but PORT0 device always stays as mouse for both game. But I can actually select my two controllers on the drop down and replace the mouse by the second one, and it works in game after that"

Now on start of the game, the two controllers are well configured and there is no need to go in GUI to modify the one in PORT0, it is already set to one of the joystick, as is PORT1

But sadly, keeping the same `hotsprefs.conf`, if only one controller is connected, this controller is not selected anymore in the GUI, you only see port 0: mouse and port1: keyboard as joystick.
You can still reassociate it, but it is not there by default, and you can't even enter GUI with it or exit game, you need the keyboard

The issue with the missing custom configuration for hotkey and associated buttons is still there too, but I will maybe create a separate issue for that as it seems it is a specific one, also appearing for ADF

**Comment by HoraceAndTheSpider** [comment]
If the game is one player , port 0 will likely default to mouse: this is a more natural Amiga behaviour. 

If there is a 2 player game that has the mouse set, and shouldn’t, check it is correct in the xml.

If the xml is wrong, please report it on the xml builds github and we will fix it :)

**Comment by HoraceAndTheSpider** [comment]
You should make sure you don’t have “redundantl” joy2 etc settings in the hostprefs - I.e only use controller_1=joy2 if you have multiple controllers connected. 

Otherwise it attempts to connect a controller which isn’t there, and can reset the other inputs you’ve already set

**Comment by Voljega** [comment]
So ideally the hostprefs file should also be custom generated based on how many controllers are currently connected before launching each game ?

**Comment by HoraceAndTheSpider** [comment]
> So ideally the hostprefs file should also be custom generated based on how many controllers are currently connected before launching each game ?

The hostprefs should reflect whatever is available on the specific host machine. Doesnt need to be before every game - most users are not changing the number of controllers they have connected.

**Comment by midwan** [comment]
@Voljega 
The `hostprefs.conf` file is now obsolete - all relevant options have been moved to `amiberry.conf` instead, which includes several defaults. Unless a config file (or the WHDLoad XML) overrides them, the defaults will be used instead.

You can check the wiki for a full documentation of what's available: https://github.com/midwan/amiberry/wiki/Amiberry.conf-options
